### PR TITLE
Adding the AWS-RunAnsible SSM Document

### PR DESCRIPTION
**Pull request by silvax**
Issue #, if available: None

Description of changes: Adding AWS-RunAnsible documet. This SSM Document allows to run Ansible automation using complex playbooks. It also supports verbose level. The playbooks can be supplied via a web URL, S3 location or in a Zip file. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    - 'elif cat /etc/issue|grep -i ''\\S'' ; then '
+    - '   sudo yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm'
+    - '   sudo yum -y update'
+    - '   sudo yum -y install python-pip'
+    - '   sudo pip install ansible --upgrade'
+    - '   sudo yum -y install wget'
+    - '   sudo yum -y install unzip'
+    - '   sudo pip install awscli --upgrade'
+    - else
+    - '   echo "There was a problem installin or updating the required tools for the
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Typo: installing

File: Documents/Command/AWS-RunAnsible/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Configure SSM Agent CloudWatch Logging
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/README.md]: This seems a wrong title

File: Documents/Command/AWS-RunAnsible/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Supported for *Linux*
+
+## Supported SSM Agent Versions
+
+Agent Version 2.0.902.0 and above
+
+## Parameters
+
+### playbook
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/README.md]: Section titles should start with an uppercase letter

File: Documents/Command/AWS-RunAnsible/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+### PlaybookURL
+
+If you don''t specify playbook YAML, then you must specify a URL where the playbook is stored. This url can refer to a playbook yaml file, an s3 prefix, or an archive (TAR GZIP/BZIP2, Zip) containing all your playbook artifacts. You can specify the URL in the following formats: http://example.com/playbook.yml or s3://examplebucket/playbook.url. You can also specify an s3 URL for a directory containing the playbooks like s3://mybucket/myplabooks/.  For security reasons, you can''t specify a URL with quotes.
+
+### PlaybookFile
+
+The playbook file to run (including relative path), when specifying a playbook bundle like a zip file or an s3 path to a folder in a bucket
+
+### extravars
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/README.md]: Section titles should start with an uppercase letter

File: Documents/Command/AWS-RunAnsible/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+Additional variables to pass to Ansible at runtime. Enter a space separated list of key/value pairs. For example: color=red flavor=lime
+
+### WorkingDirectory
+
+Current working directory to use for execution; if not specified, use random temp directory.
+
+### check
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/README.md]: Section titles should start with an uppercase letter

File: Documents/Command/AWS-RunAnsible/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+Use the check parameter to perform a dry run of the Ansible execution
+
+### verbose
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/README.md]: Section titles should start with an uppercase letter

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+  is executed. This version also supports zip files via http or s3 URLs. It also supports
+  providing an s3 URL for a directory containing the playbook files. '
+parameters:
+  playbook:
+    type: String
+    description: (Optional) If you don't specify a URL, then you must specify playbook
+      YAML in this field.
+    default: ''
+    displayType: textarea
+  PlaybookURL:
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: The current published version (https://console.aws.amazon.com/systems-manager/documents/AWS-RunAnsiblePlaybook/content?region=us-east-1) uses parameter name "playbookurl". Let's keep the same for backward compatibility.

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    description: '(Optional) If you don''t specify playbook YAML, then you must specify
+      a URL where the playbook is stored. This url can refer to a playbook yaml file,
+      an s3 prefix, or an archive (TAR GZIP/BZIP2, Zip) containing all your playbook
+      artifacts. You can specify the URL in the following formats: http://example.com/playbook.yml
+      or s3://examplebucket/playbook.url. You can also specify an s3 URL for a directory
+      containing the playbooks like s3://mybucket/myplabooks/.  For security reasons,
+      you can''t specify a URL with quotes.'
+    default: ''
+    allowedPattern: ^\s*$|^(http|https|s3)://[^']*$
+  PlaybookFile:
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Let's stick to the same parameter naming convention (all lowercase)

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    type: String
+    description: (Optional) The playbook file to run (including relative path), when
+      specifying a playbook bundle like a zip file or an s3 path to a folder in a
+      bucket
+    default: ssm-playbook.yml
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Can we add input validation, i.e. an allowedPatterns that ensures a .yml file is provided?

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  extravars:
+    type: String
+    description: '(Optional) Additional variables to pass to Ansible at runtime. Enter
+      a space separated list of key/value pairs. For example: color=red flavor=lime'
+    default: SSM=True
+    displayType: textarea
+    allowedPattern: ^$|^\w+\=[^\s|:();&]+(\s\w+\=[^\s|:();&]+)*$
+  WorkingDirectory:
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Let's stick to the same parameter naming convention (all lowercase)

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    default: ''
+  check:
+    type: String
+    description: ' (Optional) Use the check parameter to perform a dry run of the
+      Ansible execution.'
+    allowedValues:
+    - 'True'
+    - 'False'
+    default: 'False'
+  verbose:
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Can we have an option to disable verbosity to keep the existing behavior?

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    - -vvv
+    - -vvvv
+    default: -v
+mainSteps:
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    runCommand:
+    - '#!/bin/bash'
+    - 'echo "Installing and or updating required tools: Ansible, wget unzip ...."
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: This is a change in scope for the document. The document verb ("Run") suggest you will be running an Ansible playbook, but instead the document installs and runs ansible.
Let's not mix the two. One document should do 1 action. Ansible installation should be done outside of this RunCommand document.

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    - '#   PlaybookFile="ssm-playbook.yml"'
+    - '#else'
+    - PlaybookFile="{{PlaybookFile}}"
+    - '#fi'
+    - if [ -z "{{playbook}}" ] ; then
+    - '   PlaybookURL="{{PlaybookURL}}"'
+    - '   filename=${PlaybookURL##*/}'
+    - '   extension=${filename##*.}'
+    - '   if [[ "${PlaybookURL}" == http* ]]; then'
+    - '   if [[ "{{PlaybookURL}}" == */ ]]; then'
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Can we catch this invalid value in the allowedPattern of the PlaybookURL parameter somehow?

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    - '     echo "http URL is only supportted for compressed files." >&2'
+    - '     exit 1'
+    - '   fi'
+    - '#-------------- added for zip support---------------'
+    - '# If we have an extensionension, assume we have a single file and copy it;
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: Extension is mispelled

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    - '     esac'
+    - '  fi'
+    - '#---------------added for zip support --------------'
+    - '  if [ $? -ne 0 ]; then'
+    - '       echo "There was a problem downloading the playbook. Make sure the URL
+      is correct and that the playbook exists." >&2'
+    - '       exit 1'
+    - '   fi'
+    - '  elif [[ "${PlaybookURL}" == s3* ]] ; then'
+    - '# If we have an extensionension, assume we have a single file and copy it;
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: The original code runs an "aws --version" and checks the $? value to ensure the AWS CLI is installed.

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+    - '        fi'
+    - '        ;;'
+    - '      *) echo "Unsupported compression format"'
+    - '        exit 1'
+    - '        ;;'
+    - '     esac'
+    - '  fi'
+    - else
+    - '   if [ -z "{{PlaybookURL}}" ] ; then'
+    - '     echo "The playbook URL is not is not set and no playbook YAML was provided.
```
Comment (alemartini) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: "is not" is used twice

File: Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+mainSteps:
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    runCommand:
+    - '#!/bin/bash'
+    - 'echo "Installing and or updating required tools: Ansible, wget unzip ...."
+      >&2'
+    - 'if cat /etc/issue|grep -i Amazon ; then '
+    - '   sudo pip install ansible --upgrade'
```
Comment (Drudenhaus) [inline Documents/Command/AWS-RunAnsible/AWS-RunAnsible.yaml]: The installed version of pip and any other executable whose name changes with different versions should be detected. For example, Ubuntu 16.04 comes with (only) Python 3.5 by default and the pip executable is "pip3" not "pip". Here's an example you can (test and) use:

```bash
#!/bin/bash
if command -v pip &> /dev/null; then
    PIP=pip
elif command -v pip2 &> /dev/null; then
    PIP=pip2
elif command -v pip3 &> /dev/null; then
    PIP=pip3
else
    echo "Failed to find Python package manager (pip)!"
    exit -1
fi
echo "pip = '$PIP'"
```

Related to [issue 26](https://github.com/awslabs/aws-systems-manager/issues/26) regarding unsafe Python version assumptions.

**Comment by silvax** [comment]
Removing until we figure out the right approach for this

**Pull request closed without merging**